File: Data Structures/src/hard/Shawn and the universe.swift
import Foundation

let MAXN = 1000010

class Node {
    var cnt: Int
    var left: Node!
    var right: Node!

    init(_ cnt: Int, _ left: Node?, _ right: Node?) {
        self.cnt = cnt
        self.left = left
        self.right = right
    }

    func insert(_ s: Int, _ e: Int, _ v: Int) -> Node {
        if v < s || v > e {
            return self
        }

        if s == e {
            return Node(self.cnt + 1, nullNode, nullNode)
        }

        let mid = (s + e) >> 1

        return Node(
            self.cnt + 1,
            self.left.insert(s, mid, v),
            self.right.insert(mid + 1, e, v)
        )
    }
}

var nullNode = Node(0, nil, nil)
var root = Array(repeating: Node(0, nil, nil), count: MAXN)
var arr = Array(repeating: 0, count: MAXN)

var k = 0

func query(_ a: Node, _ b: Node, _ s: Int, _ e: Int) -> Int {
    let chk1 = b.cnt - a.cnt
    let chk2 = e - s + 1

    if chk1 == chk2 || e <= k {
        return -1
    }

    if s == e {
        return s
    }

    let mid = (s + e) >> 1

    var ret = query(a.left, b.left, s, mid)
    if ret == -1 {
        ret = query(a.right, b.right, mid + 1, e)
    }

    return ret
}

// Input
let first = readLine()!.split(separator: " ").map { Int($0)! }
let n = first[0]
k = first[1]

let values = readLine()!.split(separator: " ").map { Int($0)! }

for i in 0..<n {
    arr[i] = values[i]
}

nullNode.left = nullNode
nullNode.right = nullNode

root[0] = nullNode.insert(1, 1000001, arr[0])

for i in 1..<n {
    root[i] = root[i - 1].insert(1, 1000001, arr[i])
}

let q = Int(readLine()!)!

for _ in 0..<q {
    let lr = readLine()!.split(separator: " ").map { Int($0)! }
    var l = lr[0] - 1
    var r = lr[1] - 1

    let res = query(
        l == 0 ? nullNode : root[l - 1],
        root[r],
        1,
        1000001
    )

    print(res)
}

File: Data Structures/src/hard/Shawn and the universe.java
import java.io.*;
import java.util.*;

public class Main {

    static final int MAXN = 1000010;

    static int n, q, k;
    static int[] arr = new int[MAXN];
    static Node[] root = new Node[MAXN];

    static class Node {
        int cnt;
        Node left, right;

        Node(int cnt, Node left, Node right) {
            this.cnt = cnt;
            this.left = left;
            this.right = right;
        }

        Node insert(int s, int e, int v) {
            if (v < s || v > e) return this;

            if (s == e) {
                return new Node(this.cnt + 1, nullNode, nullNode);
            }

            int mid = (s + e) >> 1;

            return new Node(
                this.cnt + 1,
                this.left.insert(s, mid, v),
                this.right.insert(mid + 1, e, v)
            );
        }
    }

    static Node nullNode = new Node(0, null, null);

    static int query(Node a, Node b, int s, int e) {
        int chk1 = b.cnt - a.cnt;
        int chk2 = e - s + 1;

        if (chk1 == chk2 || e <= k) return -1;

        if (s == e) return s;

        int mid = (s + e) >> 1;

        int ret = query(a.left, b.left, s, mid);
        if (ret == -1) {
            ret = query(a.right, b.right, mid + 1, e);
        }

        return ret;
    }

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner(System.in);

        n = fs.nextInt();
        k = fs.nextInt();

        for (int i = 0; i < n; i++) {
            arr[i] = fs.nextInt();
        }

        nullNode.left = nullNode;
        nullNode.right = nullNode;

        root[0] = nullNode.insert(1, 1000001, arr[0]);

        for (int i = 1; i < n; i++) {
            root[i] = root[i - 1].insert(1, 1000001, arr[i]);
        }

        q = fs.nextInt();

        StringBuilder sb = new StringBuilder();

        while (q-- > 0) {
            int l = fs.nextInt();
            int r = fs.nextInt();

            l--; r--;

            int res = query(
                (l == 0 ? nullNode : root[l - 1]),
                root[r],
                1,
                1000001
            );

            sb.append(res).append('\n');
        }

        System.out.print(sb);
    }

    // Fast I/O
    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        FastScanner(InputStream is) {
            br = new BufferedReader(new InputStreamReader(is));
        }

        int nextInt() throws IOException {
            while (st == null || !st.hasMoreElements()) {
                st = new StringTokenizer(br.readLine());
            }
            return Integer.parseInt(st.nextToken());
        }
    }
}
